File: 015.NN_UIViewAnimationPractice/NN_UIViewAnimationPractice/UIViewAnimationPractice/Controllers/NNBackgroundAnimationViewController.m
//
//  NNBackgroundAnimationViewController.m
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2018/8/11.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNBackgroundAnimationViewController.h"

@interface NNBackgroundAnimationViewController ()

@end

@implementation NNBackgroundAnimationViewController

- (NSArray *)buttonTitleArray {
    return @[@"Alpha"];
}

#pragma mark - 处理事件区域
- (void)buttonClick:(UIButton *)sender {
    [self changeBackground];
}

/**
 *  relativeStartTime 动画在什么时候开始
 *  relativeDuration 动画所持续的时间
 */
- (void)changeBackground {
    [UIView animateKeyframesWithDuration:10.0 delay:0.f options:UIViewKeyframeAnimationOptionCalculationModeLinear animations:^{
        [UIView addKeyframeWithRelativeStartTime:0.f relativeDuration:1 / 5.0 animations:^{
            self.nnView.backgroundColor = [UIColor blueColor];
        }];
        [UIView addKeyframeWithRelativeStartTime:1 / 5.0 relativeDuration:1 / 5.0 animations:^{
            self.nnView.backgroundColor = [UIColor yellowColor];
        }];
        [UIView addKeyframeWithRelativeStartTime:2 / 5.0 relativeDuration:1 / 5.0 animations:^{
            self.nnView.backgroundColor = [UIColor redColor];
        }];
        [UIView addKeyframeWithRelativeStartTime:3 / 5.0 relativeDuration:1 / 5.0 animations:^{
            self.nnView.backgroundColor = [UIColor orangeColor];
        }];
        [UIView addKeyframeWithRelativeStartTime:4 / 5.0 relativeDuration:1 / 5.0 animations:^{
            self.nnView.backgroundColor = [UIColor whiteColor];
        }];
    } completion:^(BOOL finished) {
        self.nnView.backgroundColor = [UIColor cyanColor];
        NSLog(@"动画结束");
    }];
}

@end

File: 017.NN_UIViewAnimationPractice_swift/NN_UIViewAnimationPractice_swift/UIViewAnimationPractice/Controllers/NNBackgroundAnimationViewController.swift
//
//  NNBackgroundAnimationViewController.swift
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

class NNBackgroundAnimationViewController: NNBaseViewController {

    override func buttonClick(index: NSInteger) {
        
        // 按钮点击事件
        UIView.animateKeyframes(withDuration: 10.0, delay: 0, options: UIView.KeyframeAnimationOptions.calculationModeLinear, animations: {
            UIView.addKeyframe(withRelativeStartTime: 0, relativeDuration: 1.0 / 5.0, animations: {
                self.nnView.backgroundColor = UIColor.blue
            })
            UIView.addKeyframe(withRelativeStartTime: 1.0/5.0, relativeDuration: 1.0/5.0, animations: {
                self.nnView.backgroundColor = UIColor.yellow
            })
            UIView.addKeyframe(withRelativeStartTime: 2.0/5.0, relativeDuration: 1.0/5.0, animations: {
                self.nnView.backgroundColor = UIColor.red
            })
            UIView.addKeyframe(withRelativeStartTime: 3.0/5.0, relativeDuration: 1.0/5.0, animations: {
                self.nnView.backgroundColor = UIColor.orange
            })
            UIView.addKeyframe(withRelativeStartTime: 4.0/5.0, relativeDuration: 1.0/5.0, animations: {
                self.nnView.backgroundColor = UIColor.white
            })
            
        }) { (finished) in
            self.nnView.backgroundColor = UIColor.cyan
        }
    }

    // 数组元素
    override func buttonTitleArrayMethod() -> NSArray {
        return ["BackgroundColor"]
    }
    
}
